### ISSUE TITLE
Curreent SyncBatchNorm (a synchronized implementation) slow down DDP training

### ISSUE THREAD

**Issue by gzygzy9211** [pytorch/pytorch#36452]
## 🚀 Feature

An asynchronized implementation w.r.t to the host thread

## Motivation

Current implementation of ``SyncBatchNorm`` is not only synchronized with other workers but also block the host thread, so the host cannot continue issue computing task to the GPU like it does with ``Module`` without ``SyncBatchNorm``. So the GPU cannot be saturated during training. 

After I do some research on the current implementation, I find there are 2 draw backs:
1. ``AllGather`` must be Synchronized with host thread to guarantee the all data is gathered. This is the main reason that host thread must be blocked in current implementation. Code is at [here](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/_functions.py#L33-L35)

2. The last argument of  ``batch_norm_gather_stats_with_counts``, a.k.a. ``counts`` must be a ``IntArrayRef``, but later is converted back into ``CUDATensor`` in [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Normalization.cu#L136-L137), which cause unnecessary ``MemcpyDtoH`` and ``MemcpyHtoD`` and also block the host thread. 

Here are timelines of NVVP of DPP training with and without ``SyncBatchNorm``.
![with_sync_bn](https://user-images.githubusercontent.com/2868493/79064328-91db3b00-7cda-11ea-952d-4940ed80dab7.png)
![without_sync_bn](https://user-images.githubusercontent.com/2868493/79064331-956ec200-7cda-11ea-9076-78ea3e47b829.png)
With ``SyncBatchNorm``, forward pass is abount 700ms
Without ``SyncBatchNorm``, forward pass is about 100ms

## Pitch

1. NCCL colloective functions may be submit to task sequence of a stream just like other CUDA operation, so host thread does not have to be blocked in SyncBatchNorm and other synchronization operation like gradient reduction.
2.  The last argument of  ``batch_norm_gather_stats_with_counts``, a.k.a. ``counts`` should accept ``CUDATensor`` rather than ``IntArrayRef`` or list of ``int`` at Python layer.

## Additional context

cc @VitalyFedyunin @ngimel

**Comment by gzygzy9211** [comment]
After some further research, I found that 

> ``AllGather`` must be Synchronized with host thread to guarantee the all data is gathered. This is the main reason that host thread must be blocked in current implementation. Code is at [here](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/_functions.py#L33-L35)

is not true according to [this](https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/ProcessGroupNCCL.hpp#L48-L50
). That is ``wait()`` for NCCLProcessGroup actually enqueue an operation to current stream that wait for operations in NCCL communication stream, and will not block host thread.

Hence the **ONLY REASON** that host thread is blocked is the unnecessary ``MemcpyDtoH`` and ``MemcpyHtoD`` caused by argument ``counts`` being of type ``IntArrayRef``, since they cannot be async ops under PyTorch.

**Comment by xwang233** [comment]
Closing as fixed.
